### DISCUSSION TITLE
Albedo results

### DISCUSSION THREAD

**Discussion by mandresm** [FESOM/fesom2#426] [General]
Hi everyone!

I've been running some year-long tests in Albedo with FESOM with the aim of finding which compiler and compiler flags work best in Albedo right now. If you'd like me to run more tests, discuss the results or you have suggestions please add a comment to this discussion. The results are summarized in the table below where the rows indicate the number of MPI tasks (all tests without multithreading and no OpenMP) and the column names stand for:
```
intel: standard FESOM intel compiler flags                                    -O3 -r8 -i4 -fp-model precise -no-prec-div -no-prec-sqrt -fimf-use-svml -ip -init=zero -no-wrap-margin -xHost
intel_AVX2:  same compiler flags as in Levante                                -O3 -r8 -i4 -fp-model precise -no-prec-div -no-prec-sqrt -fimf-use-svml -ip -init=zero -no-wrap-margin -march=core-avx2 -mtune=core-avx2
intel_NEC: NEC's tests flags (as in the `refactoring` branch of FESOM)        -O3 -r8 -i4 -fp-model precise -no-prec-div -no-prec-sqrt -fimf-use-svml -ip -init=zero -no-wrap-margin -march=core-avx2 -O3 -fPIC -qopt-malloc-options=2 -qopt-prefetch=5 -unroll-aggressive

gcc: standard FESOM gcc compiler flags                                        -O3 -DNDEBUG -O3 -O2 -g -ffloat-store -finit-local-zero -finline-functions -fimplicit-none -fdefault-real-8 -ffree-line-length-none -fallow-argument-mismatch
gcc_AVX2:  adding -march and -mtune                                           -O3 -DNDEBUG -O3 -O2 -g -ffloat-store -finit-local-zero -finline-functions -fimplicit-none -fdefault-real-8 -ffree-line-length-none -march=core-avx2 -mtune=core-avx2 -fallow-argument-mismatch
gcc_NEC: adding as many of the flags  provided by the NEC tests as possible   -O3 -DNDEBUG -O3 -O3 -g -ffloat-store -finit-local-zero -finline-functions -fimplicit-none -fdefault-real-8 -ffree-line-length-none -march=core-avx2 -fPIC -unroll-aggressive -fallow-argument-mismatch
```

|                | intel           | gcc             | intel_AVX2       | gcc_AVX2         | intel_NEC       | gcc_NEC         | 
|--|--|--|--|--|--|--|
|**128**             | 111m43.309s     | 115m31.859s     | 49m50.963s      | 104m33.633s     | 50m48.135s      | 97m6.001s       | 
|**512**             | 39m49.813s      | 39m27.205s      | 15m14.597s      | 35m54.552s      | 21m47.681s      | 30m26.279s      | 
|**1024**            | 30m46.163s      | 31m7.212s       | 13m27.178s      | 26m42.096s      | 19m57.458s      | 26m20.768s      | 

- Comparing the columns with flags specific to the chip architecture, one can see that in average intel compilers are yielding in average 1.7-fold performance (with a variation between 2.12 to 1.3)
- The best performance is obtained with `intel_AVX2`, so the configuration that tries to imitate Levante's one
- FESOM seems to not be scaling so well in the range between 512-1024 in Albedo
- I would take all this with a grain of salt, since we know Albedo is having I/O problems and I only ran one test per cell of that table

Notes
------
- I don't know where the `-O3 -DNDEBUG -O3` flags for the `gcc` compilations come from (@hegish any ideas?). The `gcc` tests were run with the pristine `CMakeLists.txt` of FESOM while for `gcc_AVX2` and `gcc_NEC` I have modified this line on the same file accordingly: https://github.com/FESOM/fesom2/blob/09de73060229f5eb9dcb7c378a2dc1abfb4cc499/src/CMakeLists.txt#L164
- I'm not sure if it makes sense to make these tests with the I/O problems. I opted for year-long simulations hoping that I/O problems will be masked-out, can someone from FESOM (@patrickscholz?) comment on this?
- My knowledge of GNU compiler flags is pretty limited, I tried to adapt the flags I knew work for the intel compiler to gcc, but the small improvements in performance with gcc + `-march=core-avx2` probably means that I am not doing a good job with gcc compiler flags.

Modules and libraries
----------------------
I have tried to reproduce exactly the `intel`/`openmpi` configuration for which we obtained a good performance in Levante (`intel_AVX2`), and for that I have compiled locally (using `spack`) the libraries that FESOM needs, mimicking Levante's ones.

**intel**
```
module load /albedo/home/mandresm/.spack/modules/intel-oneapi-compilers-2022.0.1-gcc-12.1.0-mech5is
module load /albedo/home/mandresm/.spack/modules/intel-oneapi-mkl-2022.0.1-gcc-12.1.0-ebd776q
module load /albedo/home/mandresm/.spack/modules/openmpi-4.1.3-intel-2021.5.0-6xzp4ab
module load /albedo/home/mandresm/.spack/modules/hdf5-1.12.2-intel-2021.5.0-xy5vq7s
module load /albedo/home/mandresm/.spack/modules/netcdf-c-4.8.1-intel-2021.5.0-ffhyl6p
module load /albedo/home/mandresm/.spack/modules/netcdf-fortran-4.5.4-intel-2021.5.0-w4vv7su
```

I have also done the equivalent for `gcc`:
**gcc**
```
module load gcc/12.1.0
module load /albedo/home/mandresm/.spack/modules/intel-oneapi-mkl-2022.0.1-gcc-12.1.0-ebd776q
module load /albedo/soft/modules/spack-modules/openmpi/4.1.3-gcc12.1.0-npib
module load /albedo/home/mandresm/.spack/modules/hdf5-1.12.2-gcc-12.1.0-7bogsh7
module load /albedo/home/mandresm/.spack/modules/netcdf-c-4.8.1-gcc-12.1.0-2u6p2ge
module load /albedo/home/mandresm/.spack/modules/netcdf-fortran-4.5.4-gcc-12.1.0-2gawmpc
module load /albedo/home/mandresm/.spack/modules/netcdf-cxx-4.2-gcc-12.1.0-codiq6i

```

Additional details
-----------------
Partition: `mpp`
FESOM branch: `refactoring`
FESOM git SHA: `40fbbd6`

Libraries compilation with spack
--------------------------------
Here is how I compiled the FESOM dependencies:

**intel**
```
spack install intel-oneapi-compilers@2022.0.1 %gcc@12.1.0
spack install openmpi@4.1.3 %intel@2021.5.0 schedulers=slurm
spack install hdf5+fortran+hl ^openmpi@4.1.3%intel@2021.5.0 schedulers=slurm
spack install netcdf-fortran ^openmpi@4.1.3%intel@2021.5.0 schedulers=slurm ^hdf5+fortran+hl
spack install netcdf-cxx ^openmpi@4.1.3%intel@2021.5.0 schedulers=slurm ^hdf5+fortran+hl
```

**gcc**
```
spack install openmpi@4.1.3 %gcc@12.1.0 schedulers=slurm
spack install hdf5+fortran+hl ^openmpi@4.1.3%gcc@12.1.0 schedulers=slurm
spack install netcdf-fortran ^openmpi@4.1.3%gcc@12.1.0 schedulers=slurm ^hdf5+fortran+hl
spack install netcdf-cxx ^openmpi@4.1.3%gcc@12.1.0 schedulers=slurm ^hdf5+fortran+hl
```

FESOM sources and binaries
-----------------------------
- `/albedo/work/user/mandresm/model_codes/fesom-2.1-refact_gcc`
- `/albedo/work/user/mandresm/model_codes/fesom-2.1-refact_gcc_AV2`
- `/albedo/work/user/mandresm/model_codes/fesom-2.1-refact_gcc_NEC`
- `/albedo/work/user/mandresm/model_codes/fesom-2.1-refact_intel`
- `/albedo/work/user/mandresm/model_codes/fesom-2.1-refact_intel_AV2`
- `/albedo/work/user/mandresm/model_codes/fesom-2.1-refact_intel_NEC`

Work directoriers for the simulations
------------------------------------
- `/albedo/work/user/mandresm/fesom-refact_gcc_1024/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_gcc_128/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_gcc_512/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_gcc_AV2_1024/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_gcc_AV2_128/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_gcc_AV2_512/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_gcc_NEC_1024/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_gcc_NEC_128/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_gcc_NEC_512/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_intel_1024/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_intel_128/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_intel_512/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_intel_AV2_1024/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_intel_AV2_128/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_intel_AV2_512/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_intel_NEC_1024/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_intel_NEC_128/run_20010101-20011231/work`
- `/albedo/work/user/mandresm/fesom-refact_intel_NEC_512/run_20010101-20011231/work`

Slurm stdout/stderr in:
- `/albedo/work/user/mandresm/fesom-refact_gcc_1024/log/fesom-refact_gcc_1024_fesom_compute_20010101-20011231_*.log`
- `/albedo/work/user/mandresm/fesom-refact_gcc_128/log/fesom-refact_gcc_128_fesom_compute_20010101-20011231_*.log`
- `/albedo/work/user/mandresm/fesom-refact_gcc_512/log/fesom-refact_gcc_512_fesom_compute_20010101-20011231_*.log`
- `/albedo/work/user/mandresm/fesom-refact_gcc_AV2_1024/log/fesom-refact_gcc_AV2_1024_fesom_compute_20010101-20011231_*.log`                                                                                                                 
- `/albedo/work/user/mandresm/fesom-refact_gcc_AV2_128/log/fesom-refact_gcc_AV2_128_fesom_compute_20010101-20011231_*.log`                                                                                                                   
- `/albedo/work/user/mandresm/fesom-refact_gcc_AV2_512/log/fesom-refact_gcc_AV2_512_fesom_compute_20010101-20011231_*.log`                                                                                                                   
- `/albedo/work/user/mandresm/fesom-refact_gcc_NEC_1024/log/fesom-refact_gcc_NEC_1024_fesom_compute_20010101-20011231_*.log`                                                                                                                 
- `/albedo/work/user/mandresm/fesom-refact_gcc_NEC_128/log/fesom-refact_gcc_NEC_128_fesom_compute_20010101-20011231_*.log`                                                                                                                   
- `/albedo/work/user/mandresm/fesom-refact_gcc_NEC_512/log/fesom-refact_gcc_NEC_512_fesom_compute_20010101-20011231_*.log`                                                                                                                   
- `/albedo/work/user/mandresm/fesom-refact_intel_1024/log/fesom-refact_intel_1024_fesom_compute_20010101-20011231_*.log`                                                                                                                     
- `/albedo/work/user/mandresm/fesom-refact_intel_128/log/fesom-refact_intel_128_fesom_compute_20010101-20011231_*.log`
- `/albedo/work/user/mandresm/fesom-refact_intel_512/log/fesom-refact_intel_512_fesom_compute_20010101-20011231_*.log`
- `/albedo/work/user/mandresm/fesom-refact_intel_AV2_1024/log/fesom-refact_intel_AV2_1024_fesom_compute_20010101-20011231_*.log`                                                                                                             
- `/albedo/work/user/mandresm/fesom-refact_intel_AV2_128/log/fesom-refact_intel_AV2_128_fesom_compute_20010101-20011231_*.log`                                                                                                               
- `/albedo/work/user/mandresm/fesom-refact_intel_AV2_512/log/fesom-refact_intel_AV2_512_fesom_compute_20010101-20011231_*.log`                                                                                                               
- `/albedo/work/user/mandresm/fesom-refact_intel_NEC_1024/log/fesom-refact_intel_NEC_1024_fesom_compute_20010101-20011231_*.log`                                                                                                             
- `/albedo/work/user/mandresm/fesom-refact_intel_NEC_128/log/fesom-refact_intel_NEC_128_fesom_compute_20010101-20011231_*.log`                                                                                                               
- `/albedo/work/user/mandresm/fesom-refact_intel_NEC_512/log/fesom-refact_intel_NEC_512_fesom_compute_20010101-20011231_*.log`  

@patrickscholz @pgierz @koldunovn @dsidoren

**Comment by patrickscholz** [comment]
When i did my test i also did them completely without writing output and restarts, so i commented in fesom_module.F90

https://github.com/FESOM/fesom2/blob/09de73060229f5eb9dcb7c378a2dc1abfb4cc499/src/fesom_module.F90#L378
and 
https://github.com/FESOM/fesom2/blob/09de73060229f5eb9dcb7c378a2dc1abfb4cc499/src/fesom_module.F90#L381

although there still might remain the problem of reading the forcing. Martin reported in MITgcm of significantly performance differences when reading the forcing between ollie and albedo

**Reply by mandresm** [reply]
I could easily repeat the tests without writing output and restarts. I'll do that by the end of the day.

**Comment by pgierz** [comment]
Silly question, for some of the compile-time runs, you have both `O2` and `O3` set. Unless I am mistaken, the more aggressive compile optimisation will win. Was specifying both intentional? If not, it sort of defeats the purpose of the test...

I also took the liberty of plotting your wall times so we get a nice overview, see attachment. 

```python
file_contents = """
                intel           gcc             intel_AVX2      gcc_AVX2        intel_NEC       gcc_NEC
----------------------------------------------------------------------------------------------------------------
128             111m43.309s     115m31.859s     49m50.963s      104m33.633s     50m48.135s      97m6.001s
512             39m49.813s      39m27.205s      15m14.597s      35m54.552s      21m47.681s      30m26.279s
1024            30m46.163s      31m7.212s       13m27.178s      26m42.096s      19m57.458s      26m20.768s
"""

import pandas as pd
import io
import matplotlib.pyplot as plt

# assuming the file content is in a variable called "file_contents"
df = pd.read_csv(io.StringIO(file_contents), sep=r'\s{2,}', engine='python')
df.index.name = "CPU Partitioning"
df = df.transpose()
# convert the dataframe to timedeltas
df = df.applymap(pd.to_timedelta)
df.plot()
```


![image](https://user-images.githubusercontent.com/2444231/219675239-9e14292f-639e-4ecf-b46d-271ff9e8d133.png)

And also a ping to @sebhinck who is our spack expert: this may be of interest for the overall software stack!

**Reply by pgierz** [reply]
(Sorry, the label is obviously wrong: should be "MPI Tasks" ;-) )

**Reply by mandresm** [reply]
About the first question, I have written this in the first post:
> I don't know where the -O3 -DNDEBUG -O3 flags for the gcc compilations come from (@hegish any ideas?). The gcc tests were run with the pristine CMakeLists.txt of FESOM while for gcc_AVX2 and gcc_NEC I have modified this line on the same file accordingly: https://github.com/FESOM/fesom2/blob/09de73060229f5eb9dcb7c378a2dc1abfb4cc499/src/CMakeLists.txt#L164

So my problem is that I don't know where is the first `-O3 -DNDEBUG -O3` come from.

And thanks for the nice plot!

**Comment by sebhinck** [comment]
@mandresm When did you build the OpenMPI with spack? I would recommend to only use the OpenMPI provided by us admins, because there might be some cruicial flags that you were missing. However, I have configured the OpenMPI Spack package last week, so that user would automatically build their own "optimised" version. But I don't really see the point of doing so..

What Martin Losch and I have seen is that, due to the heavily increased number of cores per node (compared to Ollie), the memory bandwidth might be a crucial factor. So you might want to decrease the number of threads per node.

Thanks @mandresm  for your tests!

**Reply by mandresm** [reply]
I built open-mpi on the 14th of February. I was trying to get as close as possible to our tests in Levante (same intel-compiler and openmpi version), that's why I built it myself, but trying to run with what you are providing is the next step for me :)

> What Martin Losch and I have seen is that, due to the heavily increased number of cores per node (compared to Ollie), the memory bandwidth might be a crucial factor. So you might want to decrease the number of threads per node.

I'm not sure I'm understanding this, you don't mean to decrease the number of MPI tasks per node, right?

**Reply by sebhinck** [reply]
Ok, on the 14th Feb I already applied the change.

The first issues that you had with building on Albedo, why you chose to build your own packages as closely to the packages on Levante, was very likely due to the fact that I was not aware of the intel-oneapi-compiler ship two compilers: intel and oneapi!
I provided all packages only for the oneapi compiler, not the "intel" compiler. Therefore, I would recommend to at least try to use the system installation again! :)
For some packages, e.g. hdf5, there are some other tricks that should improve their performance (Hint from Natalja that she had from other HPC specialists from HLRN). I never tested if this does actually impact the performance, but it won't be worse! ;)


Correct, try not to fill each core of a node. The NEC benchmarkers also reported that, because of a system process for the filesystem that needs to be running on all nodes, 127 processes per node might give better performance than 128.

**Reply by mandresm** [reply]
> Correct, try not to fill each core of a node. The NEC benchmarkers also reported that, because of a system process for the filesystem that needs to be running on all nodes, 127 processes per node might give better performance than 128.

Okay... but that would mean that for applications running with MPI and OpenMP we have to play with the distribution and cpu bindings so that one MPI task fall always in a single NUMA domain, correct?

**Reply by mandresm** [reply]
I'll run with the modules you provide by default and write down the results here. I believe in those modules there are some missing options needed for OpenIFS, but if that's the case I'll come back to you :)

**Reply by sebhinck** [reply]
> > Correct, try not to fill each core of a node. The NEC benchmarkers also reported that, because of a system process for the filesystem that needs to be running on all nodes, 127 processes per node might give better performance than 128.
> 
> Okay... but that would mean that for applications running with MPI and OpenMP we have to play with the distribution and cpu bindings so that one MPI task fall always in a single NUMA domain, correct?

Yes, this has to be done for fine-tuning anyway :-)
But Fesom is not yet hybrid, right? Or are you talking about the refactored code?

**Reply by sebhinck** [reply]
> I'll run with the modules you provide by default and write down the results here. I believe in those modules there are some missing options needed for OpenIFS, but if that's the case I'll come back to you :)

Yes please, give me feedback what is missing!

**Reply by mandresm** [reply]
> Yes, this has to be done for fine-tuning anyway :-)
But Fesom is not yet hybrid, right? Or are you talking about the refactored code?

What I had in mind when I asked that was OpenIFS, but what I am currently testing is the refactored code of FESOM, however, for the moment without OpenMP threads

**Comment by pgierz** [comment]
OK, my silly question was indeed silly, I don't know how to read apparently, that was part of Miguel's notes...

**Comment by hegish** [comment]
> I don't know where the -O3 -DNDEBUG -O3 flags for the gcc compilations come from (@hegish any ideas?)

@mandresm Did you get the flags from the final make command, or are they already visible in the CMakeLists.txt?

**Reply by mandresm** [reply]
I cannot see the `-O3 -DNDEBUG` in the CMakeLists.txt. I got those lines from `fesom-2.1-/build/src/CMakeFiles/fesom.dir/flags.make`

**Reply by mandresm** [reply]
This is how I built FESOM: mkdir -p build; cd build; cmake ..;   make install -j \`nproc --all\`

**Reply by hegish** [reply]
> This is how I built FESOM: mkdir -p build; cd build; cmake ..; make install -j `nproc --all`

Could you change the make command to
`make VERBOSE=1`
and copy the full paragraph for one of the FESOM .F90 files here?

**Reply by mandresm** [reply]
Here you go!

```
[ 62%] Building Fortran object src/CMakeFiles/fesom.dir/gen_modules_read_NetCDF.F90.o                                                                                                                                                         
cd /albedo/work/user/mandresm/model_codes/fesom-2.1/build/src && /albedo/home/mandresm/.spack/sw/openmpi/4.1.3-6xzp4ab/bin/mpif90 -DDISABLE_MULTITHREADING -DMETISRANDOMSEED=35243 -DMETIS_VERSION=5 -DPARMS -DPART_WEIGHTED -I/albedo/home/ma
ndresm/.spack/sw/netcdf-fortran/4.5.4-w4vv7su/include -I/albedo/work/user/mandresm/model_codes/fesom-2.1/lib/parms/src/../include -I/albedo/work/user/mandresm/model_codes/fesom-2.1/src/async_threads_cpp -I/albedo/work/user/mandresm/model_
codes/fesom-2.1/build/src/async_threads_cpp -O3 -r8 -i4 -fp-model precise -no-prec-div -no-prec-sqrt -fimf-use-svml -ip -init=zero -no-wrap-margin -march=core-avx2 -mtune=core-avx2 -c /albedo/work/user/mandresm/model_codes/fesom-2.1/src/MOD_ICE.F90 -o CMakeFiles/fesom.dir/MOD_ICE.F90.o                                                                                                                                                                                              cd /albedo/work/user/mandresm/model_codes/fesom-2.1/build/src && /albedo/home/mandresm/.spack/sw/openmpi/4.1.3-6xzp4ab/bin/mpif90 -DDISABLE_MULTITHREADING -DMETISRANDOMSEED=35243 -DMETIS_VERSION=5 -DPARMS -DPART_WEIGHTED -I/albedo/home/mandresm/.spack/sw/netcdf-fortran/4.5.4-w4vv7su/include -I/albedo/work/user/mandresm/model_codes/fesom-2.1/lib/parms/src/../include -I/albedo/work/user/mandresm/model_codes/fesom-2.1/src/async_threads_cpp -I/albedo/work/user/mandresm/model_
codes/fesom-2.1/build/src/async_threads_cpp -O3 -r8 -i4 -fp-model precise -no-prec-div -no-prec-sqrt -fimf-use-svml -ip -init=zero -no-wrap-margin -march=core-avx2 -mtune=core-avx2 -c /albedo/work/user/mandresm/model_codes/fesom-2.1/src/g
en_halo_exchange.F90 -o CMakeFiles/fesom.dir/gen_halo_exchange.F90.o     
```

**Comment by mandresm** [comment]
**EDIT**: No simulations crashed anymore, the results in the table are the final ones.

Following the comments from @patrickscholz (commenting out the output parts of the code), Malte's (16, 32, 64, 127, 128 tasks per node) and @sebhinck's (setting `cpus-per-task=<floor(num_cores/num_tasks_per_node)>` and exporting the `SRUN_CPUS_PER_TASK` variable)

For consistency, I have not switched yet switched to the modules provided by the Albedo admins, but that's the next step.

I have not tested yet Malte's suggestion on "how the nodes are spread all over the racks". Another thing to test soon. I did not have the time yet to use `xthi` to optimise my bindings. 

This is the new table with the findings:

nhrs: node hours
the rows numbers stand for `<total_number_of_tasks_requested>_<tasks_per_node>`

|                 | intel           | gcc             | intel_AV2       | gcc_AV2         | intel_NEC       | gcc_NEC         | 
|--|--|--|--|--|--|--|
|128_16          | 80m0.165s/10.667nhrs | 74m31.692s/9.937nhrs | 19m23.906s/2.586nhrs | 68m10.891s/9.091nhrs | 19m54.267s/2.654nhrs | 61m10.275s/8.156nhrs | 
|128_32          | 75m57.537s/5.064nhrs | 75m58.354s/5.065nhrs | 21m42.767s/1.448nhrs | 69m48.778s/4.654nhrs | 20m45.257s/1.384nhrs | 62m58.651s/4.199nhrs | 
|128_64          | 84m53.406s/2.83nhrs | 88m11.076s/2.939nhrs | 28m37.699s/0.954nhrs | 79m52.522s/2.663nhrs | 28m46.656s/0.959nhrs | 72m17.780s/2.41nhrs | 
|128_127         | 94m14.202s/1.571nhrs | 95m26.982s/1.591nhrs | 47m18.401s/0.788nhrs | 88m25.455s/1.474nhrs | 47m46.613s/0.796nhrs | 82m54.577s/1.382nhrs | 
|128_128         | 110m24.004s/1.84nhrs | 112m31.655s/1.875nhrs | 49m50.317s/0.831nhrs | 104m8.695s/1.736nhrs | 50m23.514s/0.84nhrs | 101m33.611s/1.693nhrs | 
||||||||
|512_16          | 22m17.449s/11.888nhrs | 22m51.391s/12.19nhrs | 8m6.820s/4.327nhrs | 20m58.208s/11.184nhrs | 8m44.440s/4.662nhrs | 19m19.548s/10.307nhrs | 
|512_32          | 22m28.708s/5.994nhrs | 23m19.316s/6.219nhrs | 9m8.761s/2.439nhrs | 21m32.986s/5.747nhrs | 8m59.295s/2.397nhrs | 20m20.371s/5.424nhrs | 
|512_64          | 25m41.288s/3.425nhrs | 27m5.798s/3.613nhrs | 10m26.462s/1.392nhrs | 25m18.246s/3.374nhrs | 12m24.211s/1.654nhrs | 22m51.171s/3.047nhrs | 
|512_127         | 38m6.622s/2.541nhrs | 33m43.269s/2.248nhrs | 19m12.520s/1.281nhrs | 36m3.289s/2.404nhrs | 13m46.473s/0.918nhrs | 32m35.883s/2.173nhrs | 
|512_128         | 36m40.399s/2.445nhrs | 36m33.307s/2.437nhrs | 17m7.903s/1.142nhrs | 32m47.507s/2.186nhrs | 19m47.481s/1.319nhrs | 30m21.506s/2.024nhrs | 
||||||||
|1024_16         | 13m12.074s/14.081nhrs | 33m13.446s/35.439nhrs | 6m29.514s/6.925nhrs | 23m36.656s/25.185nhrs | 9m6.592s/9.717nhrs | 15m56.607s/17.006nhrs | 
|1024_32         | 14m9.527s/7.551nhrs | 14m48.969s/7.902nhrs | 8m36.692s/4.593nhrs | 14m22.575s/7.667nhrs | 7m22.607s/3.934nhrs | 13m18.732s/7.1nhrs | 
|1024_64         | 16m19.503s/4.353nhrs | 17m50.069s/4.756nhrs | 8m46.908s/2.342nhrs | 18m17.282s/4.877nhrs | 8m8.615s/2.172nhrs | 15m22.304s/4.099nhrs | 
|1024_127        | 29m28.505s/3.93nhrs | 25m32.356s/3.405nhrs | 18m51.917s/2.515nhrs | 21m2.027s/2.805nhrs | 18m30.235s/2.467nhrs | 28m6.411s/3.748nhrs | 
|1024_128        | 28m21.548s/3.781nhrs | 27m1.034s/3.602nhrs | 17m59.465s/2.399nhrs | 25m55.957s/3.458nhrs | 15m34.870s/2.077nhrs | 25m0.063s/3.333nhrs | 

The conclusions I draw from this set of tests:

1. Intel with the architecture flags is outperforming gcc for FESOM
2. There is little difference between the `NEC` compiler flags and the `intel_AVX2` ones, the really important ones, at least for this set of tests seem to be `-march=core-avx2 -mtune=core-avx2`, without them (`intel` case in the table) things get pretty slow, as expected.
3. As @sebhinck and Malte suggested, filling the nodes with tasks is not a good idea, scaling is back after following that suggestion. Switching to 127 doesn't make a big difference. I found that 64 tasks per node is a good compromise between node hours and the running time (numbers in bold on the table) for most of the cases if we were to pick up a default for FESOM in Albedo, but I have not tried anything in between 64 and 127.
